hazelcast/src/test/java/com/hazelcast/client/alto/ClientAltoTest.java

```diff
@@ -0,0 +1,501 @@
+
+        map.put("42", "42");
+        assertEquals("42", map.get("42"));
+    }
+
+    @Test
+    public void testAltoDisabledClient_inAltoEnabledCluster() {
+        Config config = getMemberConfig();
+        Hazelcast.newHazelcastInstance(config);
+        Hazelcast.newHazelcastInstance(config);
+
+        HazelcastInstance client = HazelcastClient.newHazelcastClient();
+        IMap<String, String> map = client.getMap(randomMapName());
+
+        ClientConnectionManager connectionManager = getConnectionManager(client);
+        Collection<ClientConnection> connections = connectionManager.getActiveConnections();
+        assertEquals(2, connections.size());
+        assertNoConnectionToAltoPortsAllTheTime(connections);
+
+        map.put("42", "42");
+        assertEquals("42", map.get("42"));
+    }
+
+    @Test
+    public void testAltoClient_heartbeatsToIdleAltoChannels() {
+        Config config = getMemberConfig();
+        Hazelcast.newHazelcastInstance(config);
+
+        ClientConfig clientConfig = getClientConfig();
+        clientConfig.setProperty(HEARTBEAT_INTERVAL.getName(), "1000");
+
+        HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig);
+
+        ClientConnectionManager connectionManager = getConnectionManager(client);
+        Collection<ClientConnection> connections = connectionManager.getActiveConnections();
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+
+        ClientConnection connection = connectionManager.getRandomConnection();
+        Channel[] altoChannels = connection.getAltoChannels();
+        assertNotNull(altoChannels);
+
+        long now = System.currentTimeMillis();
+        assertTrueEventually(() -> {
+            for (Channel channel : altoChannels) {
+                assertTrue(channel.lastWriteTimeMillis() > now);
+            }
+        });
+    }
+
+    @Test
+    public void testAltoClient_heartbeatsToNotRespondingAltoChannelsTimeouts() {
+        Config config = getMemberConfig();
+        Hazelcast.newHazelcastInstance(config);
+
+        ClientConfig clientConfig = getClientConfig();
+        clientConfig.setProperty(HEARTBEAT_INTERVAL.getName(), "1000");
+        clientConfig.setProperty(HEARTBEAT_TIMEOUT.getName(), "3000");
+
+        HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig);
+        ClientConnectionManager connectionManager = getConnectionManager(client);
+        Collection<ClientConnection> connections = connectionManager.getActiveConnections();
+
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+
+        ClientConnection connection = connections.iterator().next();
+        assertTrue(connection.isAlive());
+
+        // This is a long-running task that will block the Alto thread, and it
+        // should not be able to respond to ping requests
+        spawn(() -> {
```

Comment:
Partition and IO thread combines is called an Alto thread in Alto world I guess. Is there one thread per partition?

Reply:
No, one Alto thread is in charge of the possibly multiple partitions